### PR TITLE
#1146 - Ensure Dry Run Result Right Usage

### PR DESCRIPTION
**Pull request by andrepestana-aot**
Checked and did the following to ensure the right usage of the dryrun process:

- application.students.controller
	- submitApplication - changed to use values from the dryrun submission 
- designation-agreement.institutions.controller
	- submitDesignationAgreement - changed to use values from the dryrun submission 
- education-program.controller.service
	- saveProgram - ok
- institution-location.institutions.controller
	- create - ok
- report.aest.controller
	- exportReport - didn't changed validation to be done on DTO to allow dynamic data
- student.students.controller
	- create - ok
	- update - ok
- student-account-application.aest.controller
	- approveStudentAccountApplication - ok
- student-account-application.students.controller
	- create - ok
- student-appeal.students.controller
	- submitStudentAppeal - didnt changed validation to be done on DTO to allow dynamic data
- student-scholastic-standings.institutions.controller
	- saveScholasticStanding - ok
- supporting-user.supporting-users.controller
	- updateSupportingInformation - changed to use values from the dryrun submission

### CODE REVIEW

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts [hunk 1]
```diff
@@ -202,6 +197,11 @@ export class ApplicationStudentsController extends BaseController {
       );
     }
 
+    await this.applicationControllerService.offeringIntensityRestrictionCheck(
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:200]: ❤️

File: sources/packages/backend/apps/api/src/route-controllers/report/models/report.dto.ts [hunk 2]
```diff
@@ -1,17 +1,25 @@
-import { Allow } from "class-validator";
+import { IsNotEmpty } from "class-validator";
 /**
  * Filter param dynamic json.
  */
 export class ReportFilterParamAPIInDTO {
-  [columnName: string]: unknown;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/report/models/report.dto.ts:6]: This DTO must be using dynamic key name and must not use static property keys. 

The reason is that the reports that ministry view from this page is generated in a dynamic fashion and so is the report filters.

Also this is an exceptional case because of this dynamic behavior that we must do a dry run based validation instead of class validators. 

For more context on these reports please see here https://app.zenhub.com/workspaces/student-information-management-system-5fce9df5aa1b45000e937014/issues/gh/bcgov/sims/559

File: sources/packages/backend/apps/api/src/route-controllers/report/report.aest.controller.ts [hunk 3]
```diff
@@ -66,21 +64,8 @@ export class ReportAESTController extends BaseController {
     @Body() payload: ReportsFilterAPIInDTO,
     @Res() response: Response,
   ): Promise<void> {
-    const submissionResult = await this.formService.dryRunSubmission(
-      FormNames.ExportFinancialReports,
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/report/report.aest.controller.ts:70]: Please refer to the previous comment. https://github.com/bcgov/SIMS/pull/1681/files#r1087210193

I believe we must revert this.

File: sources/packages/backend/apps/api/src/route-controllers/student-appeal/models/student-appeal.dto.ts [hunk 4]
```diff
@@ -4,18 +4,32 @@ import {
...
+
 /**
  * DTO for student appeal request.
  */
 export class StudentAppealRequestAPIInDTO {
   @IsNotEmpty()
   formName: string;
   @IsDefined()
-  formData: unknown;
+  formData: StudentIncomeAppealRequest | PartnerIncomeAppealRequest;
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student-appeal/models/student-appeal.dto.ts:20]: Same here on the context of dynamic behavior. 
Also I believe there are more appeal requests to come, so IMO we must just leave it the way it was. 

Correct me if wrong @andrewsignori-aot @guru-aot

File: sources/packages/backend/apps/api/src/route-controllers/student-appeal/student-appeal.students.controller.ts [hunk 5]
```diff
@@ -111,40 +104,11 @@ export class StudentAppealStudentsController extends BaseController {
         "There is already a pending appeal for this student.",
       );
     }
-    let dryRunSubmissionResults: DryRunSubmissionResult[] = [];
-    try {
-      const dryRunPromise = payload.studentAppealRequests.map((appeal) =>
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student-appeal/student-appeal.students.controller.ts:116]: Please refer to the comment https://github.com/bcgov/SIMS/pull/1681/files#r1087213717

Student appeal can still use dry run.

File: sources/packages/backend/apps/api/src/route-controllers/supporting-user/supporting-user.supporting-users.controller.ts [hunk 6]
```diff
@@ -141,7 +141,7 @@ export class SupportingUserSupportingUsersController extends BaseController {
   ): Promise<void> {
     // Regardless of the API call is successful or not, create/update
     // the user being used to execute the request.
-    const userQuery = this.userService.syncUser(
+    const userQuery = await this.userService.syncUser(
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/route-controllers/supporting-user/supporting-user.supporting-users.controller.ts:144]: Line 144 and 154 are expected to be promise which is resolved in line 161 (Promise.all)

**Comment by dheepak-aot** [comment]
Good work @andrepestana-aot . Please take a look at the comments, most importantly on reports controller and student appeal where it we need to retain dry run validation approach.

File: sources/packages/backend/apps/api/src/route-controllers/student-appeal/models/student-appeal.dto.ts (hunk 4 above)
Comment (guru-aot) [reply sources/packages/backend/apps/api/src/route-controllers/student-appeal/models/student-appeal.dto.ts:20]: Yes you are right, currently we know these 2.

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts (hunk 1 above)
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:200]: Not part of the PR but a bit critical, if at line 174 the `getOfferingById` does not return an offering, we can throw as below:
```ts
if(!offering){
    throw new BadRequestException(
        "Selected offering id is invalid.",
    );
}
```

File: sources/packages/backend/apps/api/src/route-controllers/designation-agreement/designation-agreement.institutions.controller.ts [hunk 7]
```diff
@@ -105,9 +105,9 @@ export class DesignationAgreementInstitutionsController extends BaseController {
     const createdDesignation =
       await this.designationAgreementService.submitDesignationAgreement(
         userToken.authorizations.institutionId,
-        payload.dynamicData,
+        submissionResult.data.dynamicData,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/designation-agreement/designation-agreement.institutions.controller.ts:108]: These values are being consumed as `any`. We need to type them in some way.
Maybe we can change the `DryRunSubmissionResult` as below:
```ts
export interface DryRunSubmissionResult<T = unknown> {
  valid: boolean;
  data?: T;
  formName: string;
}
```
The `filter((location)` and the `map((location)` are treating the `location` and `any` also due to this.
If not in this PR I believe we should do one more PR right after this and maybe also change the `data` property to `resultData` to avoid the `data.data`.

@guru-aot @dheepak-aot FYI

File: sources/packages/backend/apps/api/src/route-controllers/student-appeal/models/student-appeal.dto.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/student-appeal/models/student-appeal.dto.ts:20]: In terms of dynamic data, as a general rule, the API must define types for whatever the API need to interact with. So far for these appeals (and hopefully all others to come), the API's responsibility is just to receive and save them (also validate with the `dryRun`) as they are to later also provide them as they were received to the workflow.
Short answer, please leave it as it was following what @dheepak-aot mentioned.

File: sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts [hunk 8]
```diff
@@ -59,7 +59,7 @@ export class StudentAppealService extends RecordDataModelService<StudentAppeal>
   async saveStudentAppeals(
     applicationId: number,
     userId: number,
-    studentAppealRequests: StudentAppealRequestModel[],
+    studentAppealRequests: StudentAppealRequestAPIInDTO[],
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts:62]: The idea here was to avoid the service relying on a DTO class. Is there a reason to have it changed?

**Review by andrewsignori-aot — COMMENTED** [review]
Good work chasing the different points to be changed. Please take a look at the comments.

File: sources/packages/backend/apps/api/src/route-controllers/supporting-user/supporting-user.supporting-users.controller.ts (hunk 6 above)
Comment (andrepestana-aot) [reply sources/packages/backend/apps/api/src/route-controllers/supporting-user/supporting-user.supporting-users.controller.ts:144]: Yes, you're right. Thanks. I've trying something and forgot to remove that.

File: sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts (hunk 8 above)
Comment (andrepestana-aot) [reply sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts:62]: My idea was to remove duplicate code as we have `StudentAppealRequestModel` and `StudentAppealRequestAPIInDTO`. As I reverted student appeal changes, I'm reverting this one too. Thanks!

File: sources/packages/backend/apps/api/src/route-controllers/designation-agreement/designation-agreement.institutions.controller.ts (hunk 7 above)
Comment (andrepestana-aot) [reply sources/packages/backend/apps/api/src/route-controllers/designation-agreement/designation-agreement.institutions.controller.ts]: For now I just typed the dryrun submission result. We need to talk about it. Thanks!

Comment (andrepestana-aot) [reply sources/packages/backend/apps/api/src/route-controllers/designation-agreement/designation-agreement.institutions.controller.ts]: I made the `DryRunSubmissionResult` optional generic. Then when needed/possible we can type the `.data.data` object.

**Review by dheepak-aot — APPROVED** [review]
LGTM @andrepestana-aot 👍

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts [hunk 9]
```diff
@@ -192,16 +198,22 @@ export class ApplicationStudentsController extends BaseController {
       payload.data.selectedOfferingEndDate = studyEndDate;
     }
 
-    const submissionResult = await this.formService.dryRunSubmission(
-      programYear.formName,
-      payload.data,
-    );
+    const submissionResult: DryRunSubmissionResult<ApplicationData> =
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:201]: Thank you for doing this change.
Can the `dryRunSubmission` be also adapted as below to complete the change?
```ts
async dryRunSubmission<T = any>(
    formName: string,
    data: unknown,
  ): Promise<DryRunSubmissionResult<T>>
```

File: sources/packages/backend/apps/api/src/route-controllers/designation-agreement/designation-agreement.institutions.controller.ts [hunk 10]
```diff
@@ -81,10 +82,11 @@ export class DesignationAgreementInstitutionsController extends BaseController {
       );
     }
     // Validate the dynamic data submission.
-    const submissionResult = await this.formService.dryRunSubmission(
-      FormNames.DesignationAgreementDetails,
-      payload,
-    );
+    const submissionResult: DryRunSubmissionResult =
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/designation-agreement/designation-agreement.institutions.controller.ts:85]: The explicit type here is optional. `submissionResult` is already typed.

File: sources/packages/backend/apps/api/src/route-controllers/designation-agreement/models/designation-agreement.dto.ts [hunk 11]
```diff
@@ -72,9 +78,10 @@ export class SubmitDesignationAgreementAPIInDTO {
  * be validated by the form.io dryrun validation.
  */
 export class SubmittedLocationsAPIInDTO {
-  @Allow()
+  @IsNumber()
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/designation-agreement/models/designation-agreement.dto.ts:81]: My understanding is that `@IsNumber()` is already validated by the `@IsPositive()`.
The other places along the API using `@IsPositive()` are also not used in combination with the `@IsNumber()`.

File: sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts]: Thanks for reverting it. The code should not be considered duplicated in this case. The DTO and the service model belong to different layers and they can evolve differently.
Now that we have the API service layer more dedicated to the API it is not as a big deal as it used to be but ideally, we try to avoid using API DTOs at the service level.
@dheepak-aot @guru-aot please correct me if I am not saying something accurate here.

File: sources/packages/backend/apps/api/src/route-controllers/education-program/education-program.controller.service.ts [hunk 12]
```diff
@@ -95,10 +96,11 @@ export class EducationProgramControllerService {
     auditUserId: number,
     programId?: number,
   ): Promise<EducationProgram> {
-    const submissionResult = await this.formService.dryRunSubmission(
-      FormNames.EducationProgram,
-      payload,
-    );
+    const submissionResult: DryRunSubmissionResult<SaveEducationProgram> =
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/education-program/education-program.controller.service.ts:99]: Please see https://github.com/bcgov/SIMS/pull/1681#discussion_r1089270994

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts (hunk 9 above)
Comment (andrepestana-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:201]: Yes. Thanks!

File: sources/packages/backend/apps/api/src/route-controllers/designation-agreement/models/designation-agreement.dto.ts [hunk 13]
```diff
@@ -63,7 +66,13 @@ export class UpdateDesignationAPIInDTO {
 export class SubmitDesignationAgreementAPIInDTO {
   @Allow()
   dynamicData: unknown;
-  @Allow()
+  /**
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/route-controllers/designation-agreement/models/designation-agreement.dto.ts:69]: I would recommend changing as below. I like the idea of using the form "`locations`" but I believe that would be more for the inline object types, for instance, when we create inline `options` types.
```ts
/**
   * Locations being designated.
   * Must be validated because it is not part of the dynamic data.
   */
```

File: sources/packages/backend/apps/api/src/services/form/form.service.ts [hunk 14]
```diff
@@ -88,41 +88,6 @@ export class FormService {
...
   }
 
-  /**
-   * Validate and save the data to the From IO environment generating also
-   * the absolute path for the formio generated submission.
-   * @param formName Name of the form to be saved.
-   * @param data Data to be saved.
-   * @returns Submission related information from formio submission result.
-   */
-  async submission(formName: string, data: any): Promise<SubmissionResult> {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/form/form.service.ts:98]: Thanks for removing the no longer needed code ❤️

File: sources/packages/backend/apps/api/src/types/form.ts [hunk 15]
```diff
@@ -1,14 +1,5 @@
-export interface DryRunSubmissionResult {
+export interface DryRunSubmissionResult<T = any> {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/types/form.ts:1]: In the future, in a separate effort, we can change the `any` to `unknown`. Right now this is which will bring some improvement and cause less impact.
Thanks for doing this change also.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changes, refactors, and adding the generic type to the dry run. Looks good 👍
